### PR TITLE
Docker setup

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Build ding2 from bbs.make
+FROM drush/drush:8 AS drush-build-bbs
+WORKDIR /build
+COPY bbs.make /build/
+RUN drush make /build/bbs.make /build/web
+
+# Build BBS theme
+FROM node:argon AS bbs-theme
+WORKDIR /build
+COPY themes/bbs/*.json themes/bbs/gulpfile.js ./bbs/
```
Comment (xendk) [inline Dockerfile]: What's going on here?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# Contruct php-fpm container image #
+####################################
+FROM reload/drupal-php7-fpm:5.6
+
+# Drupal/ding2
+WORKDIR /var/www/web
+COPY --from=drush-build-bbs /build/web ./
+
+# Composer install for FBS
+WORKDIR /var/www/web/profiles/ding2/modules/fbs
```
Comment (xendk) [inline Dockerfile]: This isn't needed. FBS uses xautoload now.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+RUN composer install
+
+# Composer install for ding_test
+WORKDIR /var/www/web/profiles/ding2/modules/ding_test
+RUN composer install
+
+# drupal_psr3
+WORKDIR /var/www/web/sites/all/modules/contrib/drupal_psr3
```
Comment (xendk) [inline Dockerfile]: Not needed, patched to use xautoload:
https://github.com/rvk-utd/bbs/blob/7ffbb30dc63f10d7837a30fb7e8a4298d9a650ac/bbs.make#L24

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+RUN composer install
+
+# Aleph
+WORKDIR /var/www/web/sites/all/modules/custom/aleph
+COPY ./modules/aleph/composer.* ./
+RUN composer install
+
+# BBS
+WORKDIR /var/www/web
+COPY ./composer.* ./
```
Comment (xendk) [inline Dockerfile]: Are you just building vendor here?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+RUN composer install
+
+# Make sure files and files/private exists
+WORKDIR /var/www/web
+RUN mkdir -p sites/default/files sites/default/private && chown -R www-data sites/default/files && chown -R www-data sites/default/private
+COPY docker/sample-assets/* sites/default/files/
```
Comment (xendk) [inline Dockerfile]: Why are we still doing this?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+# DDbasic theme
+COPY --from=ddbasic-theme /build/ddbasic /var/www/web/profiles/ding2/themes/ddbasic/
```
Comment (xendk) [inline Dockerfile]: Why build the themes? The containers do it on startup anyway. Just for faster startup if you have an existing image?

File: README.md [hunk 7]
```diff
@@ -1,30 +1,35 @@
...
 
-BBS modules and themes are symlinked into sites/all. When creating new
-modules `make build` will re-symlink without rebuilding all of DDB
-CMS.
+* `docker-compose up`
+* Grab a cup of coffee ☕️
+* Go to http://bbs.docker
 
-To bring up site, run:
+If you get "Service Unavailable" it's not started yet; go get some more coffee.
```
Comment (xendk) [inline README.md]: What when you've been away from the project for some time and we've, say, upgraded ding? How will you remember to rebuild the image?

File: README.md [hunk 8]
```diff
@@ -1,30 +1,35 @@
...
 
-``` shell
-make up
-```
+## Tips
+* Login with `docker-compose php drush uli`
+* npm, composer, drush, etc. should only be executed inside the container
+* If you want to make changes in Drupal or ding2, add `- './web:/var/www/web:cached'` to volumes for the php container in `docker-compose.yml`, run `docker cp bbs_php_1:/var/www/web ./web` and restart the container.
```
Comment (xendk) [inline README.md]: And that'll just work? I don't see ./web being mounted in the docker-compose?

File: docker-compose.yml [hunk 9]
```diff
@@ -1,53 +1,48 @@
...
+      - '/var/www/web/profiles/ding2/themes/ddbasic/node_modules/'
+      - '/var/www/web/sites/all/modules/custom/aleph/vendor/'
+      - '/var/www/web/sites/all/modules/custom/drupal_psr3/vendor/'
+      - '/var/www/web/sites/all/themes/bbs/node_modules/'
       - drush-cache:/root/.drush
-      - ./scripts/site-reset.sh:/root/site-reset.sh
     working_dir: /var/www/web
-    depends_on:
-      - db
-    links:
```
Comment (xendk) [inline docker-compose.yml]: When was `links` deprecated?

File: Dockerfile (hunk 6 above)
Comment (tripox) [reply Dockerfile]: Yes, for faster startup. This way you can destroy your containers and not have to wait for it to rebuild the theme. Then we can just run the watcher when the container starts.

File: README.md (hunk 7 above)
Comment (tripox) [reply README.md]: We can update the documentation to say `docker-compose up --build` instead.

File: README.md (hunk 8 above)
Comment (tripox) [reply README.md]: I says you need to add it manually. 
In this way we don't need docker-sync on Mac.
We could make a separate docker-compose.yml for Mac without it.

File: docker-compose.yml (hunk 9 above)
Comment (tripox) [reply docker-compose.yml]: That's a long time ago. :)

File: Dockerfile (hunk 1 above)
Comment (tripox) [reply Dockerfile]: We build the theme so we can make use of the docker layer cache and don't have to wait for it.

File: Dockerfile (hunk 2 above)
Comment (tripox) [reply Dockerfile]: It was last time I tried.

File: Dockerfile (hunk 3 above)
Comment (tripox) [reply Dockerfile]: It was last time I tried.

File: Dockerfile (hunk 4 above)
Comment (tripox) [reply Dockerfile]: Yes

File: Dockerfile (hunk 5 above)
Comment (tripox) [reply Dockerfile]: Are you talking about the sample-assets?
We could probably delete them, yes.

File: Dockerfile (hunk 1 above)
Comment (xendk) [reply Dockerfile]: Try again...

image

File: Dockerfile (hunk 2 above)
Comment (xendk) [reply Dockerfile]: Try again.

image

File: Dockerfile (hunk 3 above)
Comment (xendk) [reply Dockerfile]: Then try again.

File: Dockerfile (hunk 5 above)
Comment (xendk) [reply Dockerfile]: Yeah. They're not used anymore.

File: Dockerfile (hunk 6 above)
Comment (xendk) [reply Dockerfile]: Well, pre-everythinginthecontainer, you didn't have to wait for the to rebuild when you nuked your containers, so it's a bit back and forth.

File: README.md (hunk 7 above)
Comment (xendk) [reply README.md]: If you're telling people to rebuld all the time, we're back to square one?

File: README.md [hunk 10]
```diff
@@ -1,30 +1,35 @@
...
-``` shell
-drush css-generate
-```
+`docker cp bbs_php_1:/var/www/web ./web`
+
+Now you can run `docker-compose exec php xdebug-start`
+
+You need to map the following paths (local --> remote):
+* `./web --> /var/www/web`
+* `./modules --> /var/www/sites/all/modules/custom`
```
Comment (xendk) [inline README.md]: Oh, and that's not where the modules are on live.

File: README.md (hunk 7 above)
Comment (tripox) [reply README.md]: No, Docker caches layers so it only rebuilds what is needed and skips what is unchanged.

File: Dockerfile (hunk 6 above)
Comment (tripox) [reply Dockerfile]: You don't have to wait for it to rebuild when you nuke the containers.
That's what you had to do before.
Now the build is cached.

File: README.md (hunk 10 above)
Comment (tripox) [reply README.md]: We clear the cache when we spin it up.
This way it's just a lot easier to mount everything in the modules folder.

Comment (tripox) [reply README.md]: Otherwise you'd have to specify new modules.

**Comment by tripox** [comment]
We create an image for the PHP service, so we can run some drush commands before the environment is ready.

File: Dockerfile (hunk 1 above)
Comment (xendk) [reply Dockerfile]: It doesn't use the layer cache here, not according to my quick re-running.

I found it a bit odd that it's only copying json and js files into the container. How can gulp build the CSS then?

Comment (xendk) [reply Dockerfile]: Arg. Never mind about the caching. `docker-compose` chaches per default.

Comment (xendk) [reply Dockerfile]: Running gulp doesn't work as there's no SASS files. The reason it seems to work is that the gulp container bulids it on startup.

File: Dockerfile (hunk 6 above)
Comment (xendk) [reply Dockerfile]: No I didn't have to rebuild when I nuked the containers as the theme was mounted in, so the generated CSS was in the host FS. But anyhoo...

**Pull request closed without merging**